File: Frameworks/libatrac9/libatrac9/LibAtrac9/CSharp/LibAtrac9/Quantization.cs
using System;

namespace LibAtrac9
{
    internal static class Quantization
    {
        public static void DequantizeSpectra(Block block)
        {
            foreach (Channel channel in block.Channels)
            {
                Array.Clear(channel.Spectra, 0, channel.Spectra.Length);

                for (int i = 0; i < channel.CodedQuantUnits; i++)
                {
                    DequantizeQuantUnit(channel, i);
                }
            }
        }

        private static void DequantizeQuantUnit(Channel channel, int band)
        {
            int subBandIndex = Tables.QuantUnitToCoeffIndex[band];
            int subBandCount = Tables.QuantUnitToCoeffCount[band];
            double stepSize = Tables.QuantizerStepSize[channel.Precisions[band]];
            double stepSizeFine = Tables.QuantizerFineStepSize[channel.PrecisionsFine[band]];

            for (int sb = 0; sb < subBandCount; sb++)
            {
                double coarse = channel.QuantizedSpectra[subBandIndex + sb] * stepSize;
                double fine = channel.QuantizedSpectraFine[subBandIndex + sb] * stepSizeFine;
                channel.Spectra[subBandIndex + sb] = coarse + fine;
            }
        }

        public static void ScaleSpectrum(Block block)
        {
            foreach (Channel channel in block.Channels)
            {
                ScaleSpectrum(channel);
            }
        }

        private static void ScaleSpectrum(Channel channel)
        {
            int quantUnitCount = channel.Block.QuantizationUnitCount;
            double[] spectra = channel.Spectra;

            for (int i = 0; i < quantUnitCount; i++)
            {
                for (int sb = Tables.QuantUnitToCoeffIndex[i]; sb < Tables.QuantUnitToCoeffIndex[i + 1]; sb++)
                {
                    spectra[sb] *= Tables.SpectrumScale[channel.ScaleFactors[i]];
                }
            }
        }
    }
}

File: Frameworks/libatrac9/libatrac9/LibAtrac9/C/quantization.c
#include "quantization.h"
#include <string.h>
#include "tables.h"

void DequantizeSpectra(block* block)
{
	for (int i = 0; i < block->ChannelCount; i++)
	{
		channel* channel = &block->Channels[i];
		memset(channel->Spectra, 0, sizeof(channel->Spectra));

		for (int j = 0; j < channel->CodedQuantUnits; j++)
		{
			DequantizeQuantUnit(channel, j);
		}
	}
}

 void DequantizeQuantUnit(channel* channel, int band)
{
	const int subBandIndex = QuantUnitToCoeffIndex[band];
	const int subBandCount = QuantUnitToCoeffCount[band];
	const double stepSize = QuantizerStepSize[channel->Precisions[band]];
	const double stepSizeFine = QuantizerFineStepSize[channel->PrecisionsFine[band]];

	for (int sb = 0; sb < subBandCount; sb++)
	{
		const double coarse = channel->QuantizedSpectra[subBandIndex + sb] * stepSize;
		const double fine = channel->QuantizedSpectraFine[subBandIndex + sb] * stepSizeFine;
		channel->Spectra[subBandIndex + sb] = coarse + fine;
	}
}

void ScaleSpectrumBlock(block* block)
 {
	 for (int i = 0; i < block->ChannelCount; i++)
	 {
		 ScaleSpectrumChannel(&block->Channels[i]);
	 }
 }

void ScaleSpectrumChannel(channel* channel)
{
	 const int quantUnitCount = channel->Block->QuantizationUnitCount;
	 double* spectra = channel->Spectra;

	 for (int i = 0; i < quantUnitCount; i++)
	 {
		 for (int sb = QuantUnitToCoeffIndex[i]; sb < QuantUnitToCoeffIndex[i + 1]; sb++)
		 {
			 spectra[sb] *= SpectrumScale[channel->ScaleFactors[i]];
		 }
	 }
 }